src-tauri/src/github/request_client.rs

```diff
@@ -0,0 +1,427 @@
+        Ok(true)
+    }
+
+    #[allow(dead_code)]
+    pub async fn lookup_content_size(&self, url: &str) -> Result<u64, anyhow::Error> {
+        let head_response = self.send_head_request(url).await?;
+        let content_length = self.get_content_length_from_head_response(&head_response);
+        Ok(content_length)
+    }
+
+    pub async fn get_content_size_from_file(&self, path: PathBuf) -> Result<u64, anyhow::Error> {
+        let file = File::open(path).await?;
+        let metadata = file.metadata().await?;
+        Ok(metadata.len())
+    }
+
+    pub async fn download_file(
+        &self,
+        url: &str,
+        destination: &Path,
+        check_cache: bool,
+    ) -> Result<(), anyhow::Error> {
+        if check_cache {
+            //TODO (2/2) bring it back once cloudflare stops returning dynamic status
+            // self.check_if_cache_hits(url).await?;
+        }
+
+        let head_response = self.send_head_request(url).await?;
+        let head_reponse_content_length =
+            self.get_content_length_from_head_response(&head_response);
+        let head_reponse_etag = self.get_etag_from_head_response(&head_response);
+
+        let get_response: reqwest::Response = self.send_get_request(url).await?;
+        let get_reposnse_etag = self.get_etag_from_head_response(&get_response);
+
+        // Ensure the directory exists
+        if let Some(parent) = destination.parent() {
+            fs::create_dir_all(parent).await?;
+        }
+
+        // Open a file for writing
+        let mut destination_path = File::create(destination).await?;
+
+        // Stream the response body directly to the file
+        let mut stream = get_response.bytes_stream();
+        while let Some(item) = stream.next().await {
+            destination_path.write_all(&item?).await?;
+        }
+
+        let destination_file_size = self
+            .get_content_size_from_file(destination.to_path_buf())
+            .await?;
+
+        if check_cache {
+            let head_reposnse_cache_status =
+                self.get_cf_cache_status_from_head_response(&head_response);
+            info!(target: LOG_TARGET, "Cloudflare cache status: {:?}", head_reposnse_cache_status.to_str());
+        };
+
+        info!(target: LOG_TARGET, "Expected downloaded file size: {}", head_reponse_content_length);
+        info!(target: LOG_TARGET, "Downloaded file size: {}", destination_file_size);
+
+        info!(target: LOG_TARGET, "Expected etag: {}", head_reponse_etag);
+        info!(target: LOG_TARGET, "Downloaded etag: {}", get_reposnse_etag);
+
+        if head_reponse_content_length != 0 && head_reponse_content_length != destination_file_size
+        {
+            let error_message = format!(
+                "Downloaded file size does not match expected size. Expected: {}, Actual: {}",
+                head_reponse_content_length, destination_file_size
+            );
+            sentry::capture_message(&error_message, sentry::Level::Error);
```

Comment:
I only want to send it to Sentry if the entire process fails and return the reason for the failure be it checksum, size, or other failure. Any single time one of the 3 (or 9?) possible retries fails, let's not report it. Otherwise, we don't know if the new system is working. Our expectation is that there can be some strange cache anomalies and that retrying right away should help the overall success of download and extraction. If we report the failure of each attempt independently we'll continue to pollute sentry, and won't know if the end result worked or not as we'll still see all the errors in between the successes. 

If we only report after all retries are depleted we should see a significant decrease in overall failure rate, based on the idea that 1/3 will work successfully.